### ISSUE TITLE
Diverging attention kernels due to `allow_is_bidirectional_skip` branching on torch.compile

### ISSUE THREAD

**Issue by xmfan** [huggingface/transformers#44188]
### System Info

Hi, while we were updating the PyTorch transformers pin to v5.2.0, our regression tests caught a numerics issue between eager and compiled, the difference is very substantial (3.3 vs the typical e-4 accepted difference). Digging into it: https://github.com/pytorch/pytorch/pull/175274#issuecomment-3930952666, we found the cause to be in these lines (added in https://github.com/huggingface/transformers/pull/41265): 

https://github.com/huggingface/transformers/blob/147b7aa040812b079f467e777a2d2e1284167de0/src/transformers/masking_utils.py#L490-L491

We set `allow_is_bidirectional_skip=True` in a few places:
https://github.com/huggingface/transformers/blob/147b7aa040812b079f467e777a2d2e1284167de0/src/transformers/masking_utils.py#L996-L997

And in `_ignore_bidirectional_mask_sdpa`, we branch logic on whether we compile or not:
https://github.com/huggingface/transformers/blob/147b7aa040812b079f467e777a2d2e1284167de0/src/transformers/masking_utils.py#L324-L332

This issue was found on BERT but it seems like it would affect other models too.

We've also verified that removing the branching fixes the numerical difference. I'm creating this issue to ask about the best way forward here. From the PR that added it, it looks like this was necessary specifically for executorch, but the algorithm difference is also affected all other APIs that fall under `is_tracing` . Can we restrict the check?

### Who can help?

@vasqu @ArthurZucker @Cyrilvallez 

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

I believe the description is enough, but I can provide a simpler repro on request

### Expected behavior

transformers users probably shouldn't run into large numeric differences when compiling, at least not by default

**Comment by vasqu** [comment]
Just quickly commenting but essentially this caused by dropout rng being different for the different kernels (fa backend vs memory efficient). This could also happen in decoder-only models but dropout is just not really popular in modern models.

Wdyt about adding another condition on dropout? I think we have like 2-3 different names for dropout in the config so we could easily check for it (test_modeling_common should have a check for dropout somewhere surely). My issue is that we do want to avoid having to use the mask to opt into the fastest kernels where possible so see my tldr below. 

Tl;dr: I can think of 2 options
- Make dropout not skip mask creations as another condition (forces slower training)
- The test is rewritten on torch side to create their masks themself (if you pass your own 4D mask we don't do anything)

**Comment by xmfan** [comment]
I see, the options make sense, we would have to do 1 if we wanted same numerics.

But back on restricting the check approach, if the issue was with export tracing, could we be using `is_torchdynamo_exporting` instead of `is_tracing`?

https://github.com/huggingface/transformers/blob/147b7aa040812b079f467e777a2d2e1284167de0/src/transformers/masking_utils.py#L322-L325

**Comment by vasqu** [comment]
Tbh, personally I'd be for option 2 for the simple reason that users should run the fastest kernel whenever possible 😅 are there any references for the test? We know it works and we know the root cause is dropout behaving differently (different sets are disabled); not the mask being wrong.

Oh sorry overlooked the restriction: iirc, these are dyanmic control flows after the `is_tracing` check so (fullgraph) compile would complain, no? E.g. ` if mask.all(): ... else: ...`

**Comment by gowthamr-tech** [comment]
I’m working on this issue.

**Comment by i3hz** [comment]
I tested it and `padding_mask.all()` fails under `torch.compile(fullgraph=True)` .
I'm happy to put up a PR for option 1 (dropout condition) which I guess would add an argument to `_ignore_bidirectional_mask_sdpa` for dropout and a conditional check with an early return.  Should be a quick fix since the function is only called in one place .
What do you think @vasqu

**Comment by vasqu** [comment]
> personally I'd be for option 2 for the simple reason that users should run the fastest kernel whenever possible

Rereferencing myself because I honestly think it would make more sense to adjust the test on torch side: Either by modifying the test with
- Creating the mask themselves (we ignore properly prepared masks when detected, e.g. 4D mask)
- Force the specific backends to be used via https://docs.pytorch.org/docs/stable/generated/torch.nn.attention.sdpa_kernel.html

I'm hesitant here because I want users to opt into the fastest option when possible and the divergence stems from something known (dropout rng).

**Comment by GS-GOAT** [comment]
@vasqu That makes sense -I agree users should default to the fastest kernel path.
I'll hold off on this PR and wait for the decision from the torch side. Happy to close this if the consensus is that it should be addressed there instead.
That said, just wanted to note the case for this approach:-
The fix only activates when dropout > 0 (training), during model.eval(), dropout=0 so flash attention is still used with zero perf impact. The divergence users see is ~3.3, which is hard to debug without understanding SDPA backend internals. Scoping to sdpa only, keeps the blast radius minimal.

**Comment by vasqu** [comment]
Any updates @xmfan?
